src/man/DS2408.3
'\"
'\" Copyright (c) 2003-2004 Paul H Alfille, MD
'\" (paul.alfille@gmail.com)
'\"
'\" Device manual page for the OWFS -- 1-wire filesystem package
'\" Based on Dallas Semiconductor, Inc's datasheets, and trial and error.
'\"
'\" Free for all use. No waranty. None. Use at your own risk.
'\" $Id$
'\"
.TH DS2408 3  2003 "OWFS Manpage" "One-Wire File System"
.SH NAME
.TP
.B DS2408
1-Wire 8 Channel Addressable Switch
.SH SYNOPSIS
8 port switch
.PP
.B 29
[.]XXXXXXXXXXXX[XX][/[
.B latch.[0-7|ALL|BYTE]
|
.B LCD_M/[clear|home|screen|message]
|
.B LCD_H/[clear|home|yxscreen|screen|message|onoff]
|
.B LCD_H/redefchar.[0-7|ALL] LCD_H/redefchar_hex.[0-7|ALL]
|
.B PIO.[0-7|ALL|BYTE]
|
.B power
|
.B sensed.[0-7|ALL|BYTE]
|
.B strobe
|
.B por
|
.B set_alarm
|
.B out_of_testmode
|
.so man3/standard_mini.3so
]]
.SH FAMILY CODE
.PP
.I 29
.SH SPECIAL PROPERTIES
.SS latch.0 ... latch.7 latch.ALL latch.BYTE
.I read-write, binary
.br
The 8 pins (PIO) latch a bit when their state changes, either externally, or through a write to the pin. 
.br
Reading the
.I latch
property indicates that the latch has been set.
.br
Writing "true" (non-zero) to ANY 
.I latch 
will reset them all. (This is the hardware design).
.br
.I ALL
is all 
.I latch 
states, accessed simultaneously, comma separated.
.br
.I BYTE
references all channels simultaneously as a single byte. Channel 0 is bit 0.
.SS PIO.0 ...  PIO.7 PIO.ALL PIO.BYTE
.I read-write, yes-no
.br
State of the open-drain output (
.I PIO
) pin. 0 = non-conducting (off), 1 = conducting (on).
.br
Writing zero will turn off the switch, non-zero will turn on the switch. Reading the
.I PIO
state will return the switch setting. To determine the actual logic level at the switch, refer to the
.I sensed.0 ... sensed.7 sensed.ALL sensed.BYTE
property.
.br
.I ALL
references all channels simultaneously, comma separated.
.br
.I BYTE
references all channels simultaneously as a single byte. Channel 0 is bit 0.
.SS power
.I read-only, yes-no
.br
Is the
.I DS2408
powered parasitically (0) or separately on the Vcc pin (1)?
.SS sensed.0 ... sensed.7 sensed.ALL
.I read-only, yes-no
.br
Logic level at the
.I PIO
pin. 0 = ground. 1 = high (~2.4V - 5V ). Really makes sense only if the
.I PIO
state is set to zero (off), else will read zero.
.br
.I ALL
references all channels simultaneously, comma separated.
.br
.I BYTE
references all channels simultaneously as a single byte. Channel 0 is bit 0.
.SS strobe
.I read-write, yes-no
.br
RSTZ Pin Mode Control. Configures RSTZ as either RST input or STRB output:
.TP
.I 0
configured as RST input (default)
.TP
.I 1
configured as STRB output
.PP
.SS por
.I read-write, yes-no
.br
Specifies whether the device has performed power-on reset. This bit can only
be cleared to 0 under software control. As long as this bit is 1 the device
will allways respond to a conditional search.
.SS out_of_testmode
.I write-only, yes-no
.br
Write "1" to this property to make sure the device has been properly initialized on startup.
.PP
The datasheet says that under some conditions the startup (power-up) will leave the device in the "testmode" state. Any problems with "Channel Access Write" will trigger this property automaticlly, but this property makes explicit initialization possible as well.
.SS set_alarm
.I read-write, integer unsigned (0-333333333)
.br
A number consisting of 9 digits XYYYYYYYY, where:
.TP
X
select source and logical term 
.br 
.I 0
PIO   OR 
.br 
.I 1
latch OR 
.br 
.I 2
PIO   AND 
.br 
.I 3
latch AND
.TP
Y
select channel and polarity
.br 
.I 0
Unselected (LOW) 
.br
.I 1
Unselected (HIGH) 
.br
.I 2
Selected    LOW 
.br
.I 3
Selected    HIGH
.PP
All digits will be truncated to the 0-3 range. Leading zeroes are optional. Low-order digit is channel 0.
.PP
Example:
.TP
100000033
Responds on Conditional Search when latch.1 or latch.0 are set to 1.
.TP
222000000
Responds on Conditional Search when sensed.7 and sensed.6 are set to 0.
.TP
000000000 (0)
Never responds to Conditional Search.
.SH LCD_H LCD SCREEN PROPERITES
This mode uses the 
.I DS2408
attached to a Hitachi HD44780 LCD controller in 4-bit mode. See 
.I DATASHEET
for published details. Based on a commercial product from
.I HobbyBoards
by Erik Vickery.
.SS LCD_H/clear
.I write-only, yes-no
.br
This will
.I clear
the screen and place the cursor at the start.
.SS LCD_H/home
.I write-only, yes-no
.br
Positions the cursor in the
.I home
(upper left) position, but leaves the current text intact.
.SS LCD_H/screen
.I write-only, ascii text
.br
Writes to the LCD 
.I screen
at the current position.
.SS LCD_H/screenyc
.I write-only, ascii text
.br
Writes to an LCD screen at a specified location. The controller doesn't know the true LCD dimensions, but typical selections are:
2x16 2x20 4x16 and 4x20.
.TP
Y (row)
range 1 to 2 (or 4)
.TP
X (column)
range 1 to 16 (or 20)
.P
There are two formats allowed for the
.I screenyx
text, either ascii (readable text) or a binary form.
.TP 
2 binary bytes
The two first characters of the passed string have the line and row: 
e.g. "\\x02\\x04string" perl string writes "string" at line 2 column 4.
.TP 
ascii 2,12:
Two numbers giving line and row:  Separate with a comma and end with a colon
e.g. "2,4:string" writes "string" at line 2 column 4.
.TP 
ascii 12:
Single column number on the (default) first line:  End with a colon
e.g. "12:string" writes "string" at line 1 column 12.
.P
The positions are 1-based (i.e. the first position is 1,1).
.SS LCD_H/onoff
.I write-only, unsigned
.br
Sets several screen display functions. The selected choices should be added together.
.TP
4
Display on
.TP
2
Cursor on
.TP
1
Cursor blinking
.SS LCD_H/message
.I write-only, ascii text
.br
Writes a
.I message
to the LCD screen after clearing the screen first.
This is the easiest way to display a message.
.SS LCD_H/redefchar.0-7|ALL
.I write-only, binary
.br
Redefines one of 8 user-designed character glyphs for the LCD screen (5x8 pixels).
.PP
Each byte defines a horizontal line top to bottom. All 5 pixels corresponds to 0x1F and a blank line is 0x00.
.PP
Format is 8 binary bytes.
.SS LCD_H/redefchar_hex.0-7|ALL
.I write-only, ascii
.br
Redefines one of 8 user-designed character glyphs for the LCD screen (5x8 pixels).
.PP
Each byte defines a horizontal line top to bottom. All 5 pixels corresponds to 0x1F and a blank line is 0x00.
.PP
Format is 8 hexidecomal bytes (16 characters).
.SH LCD_M LCD SCREEN PROPERITES
This mode uses the 
.I DS2408
attached to a Hitachi HD44780 LCD controller in 8-bit mode. See 
.I DATASHEET
for published details. Based on a design from Maxim and a commercial product from
.I AAG.
.SS LCD_M/clear
.I write-only, yes-no
.br
This will
.I clear
the screen and place the cursor at the start.
.SS LCD_M/home
.I write-only, yes-no
.br
Positions the cursor in the
.I home
(upper left) position, but leaves the current text intact.
.SS LCD_M/screen
.I write-only, ascii text
.br
Writes to the LCD 
.I screen
at the current position.
.SS LCD_M/screenyc
.I write-only, ascii text
.br
Writes to an LCD screen at a specified location. The controller doesn't know the true LCD dimensions, but typical selections are:
2x16 2x20 4x16 and 4x20.
.TP
Y (row)
range 1 to 2 (or 4)
.TP
X (column)
range 1 to 16 (or 20)
.P
There are two formats allowed for the
.I screenyx
text, either ascii (readable text) or a binary form.
.TP 
2 binary bytes
The two first characters of the passed string have the line and row: 
e.g. "\\x02\\x04string" perl string writes "string" at line 2 column 4.
.TP 
ascii 2,12:
Two numbers giving line and row:  Separate with a comma and end with a colon
e.g. "2,4:string" writes "string" at line 2 column 4.
.TP 
ascii 12:
Single column number on the (default) first line:  End with a colon
e.g. "12:string" writes "string" at line 1 column 12.
.P
The positions are 1-based (i.e. the first position is 1,1).
.SS LCD_M/onoff
.I write-only, unsigned
.br
Sets several screen display functions. The selected choices should be added together.
.TP
4
Display on
.TP
2
Cursor on
.TP
1
Cursor blinking
.SS LCD_M/message
.I write-only, ascii text
.br
Writes a
.I message
to the LCD screen after clearing the screen first.
This is the easiest way to display a message.
.SH STANDARD PROPERTIES
.so man3/standard.3so
.SH ALARMS
Use the
.I set_alarm
property to set the alarm triggering criteria.
.SH DESCRIPTION
.so man3/description.3so
.SS DS2408
The
.B DS2408 (3)
allows control of other devices, like LEDs and relays. It extends the
.B DS2406
to 8 channels and includes memory.
.br
Alternative switches include the
.B DS2406, DS2407
and even
.B DS2450
.SH ADDRESSING
.so man3/addressing.3so
.SH DATASHEET
.br
http://pdfserv.maxim-ic.com/en/ds/DS2408.pdf
.br
http://www.hobby-boards.com/catalog/howto_lcd_driver.php
.br
http://www.maxim-ic.com/appnotes.cfm/appnote_number/3286
.SH SEE ALSO
.so man3/seealso.3so
.SH AVAILABILITY
http://www.owfs.org
.SH AUTHOR
Paul Alfille (paul.alfille@gmail.com)


NEWS
$Id$

Most everything can be found on the website:
http://www.owfs.org

Basically we support several languages:
perl, python, php, tcl

We support most of the 1-wire chips:
memmory, temperature, voltage, switches
(Actually Thermacron support is unfinished, as is DS1922, DS1923,
Java iButtons and SHA encryption is unsupported. All other 1-wire
slaves are supported).

We support serial, i2c, ethernet and USB adapters (as well as a separate backend: owserver).

We support multiple adapters and clients.

We have caching, simultaneous conversion, multithreading.

We have clients for the owserver that are light weight, platform independent,
for the shell, perl, python, Visual Basic, Java, ...

We have monitoring programs: owmon and owtap



module/swig/python/unittest/ds2408.py
#! /usr/bin/env python

"""
::BOH
$Id$
$HeadURL: http://subversion/stuff/svn/owfs/trunk/unittest/ds2408.py $

Copyright (c) 2004 Peter Kropf. All rights reserved.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
::EOH

Test suite for basic reality checks on a 1-wire network.

PIO.0/       PIO.1/     PIO.2/    PIO.3/    PIO.4/     PIO.5/
PIO.6/       PIO.7/     PIO.ALL/  PIO.BYTE/ address/   crc8/
family/      id/        latch.0/  latch.1/  latch.2/   latch.3/
latch.4/     latch.5/   latch.6/  latch.7/  latch.ALL/ latch.BYTE/
power/       present/   reset/    sensed.0/ sensed.1/  sensed.2/
sensed.3/    sensed.4/  sensed.5/ sensed.6/ sensed.7/  sensed.ALL/
sensed.BYTE/ set_alarm/ strobe/   type/
"""


import unittest
import sys
import os
import ConfigParser
import ow
import util


__version__ = '0.0'


if not os.path.exists( 'owtest.ini' ):
    raise IOError, 'owtest.ini'

config = ConfigParser.ConfigParser( )
config.read( 'owtest.ini' )

sensors = util.find_config( config, 'DS2408' )
if len( sensors ):
    load = True
else:
    load = False


class DS2408( unittest.TestCase ):
    def setUp( self ):
        ow.init( config.get( 'General', 'interface' ) )


    def testAttributes( self ):
        self.failIfEqual( len( sensors ), 0 )
        for name in sensors:
            sensor = ow.Sensor( name )
            family, id = name.split( '/' )[ -1 ].split( '.' )

            self.failUnlessEqual( family + id,                      sensor.address[ :-2 ] )
            #self.failUnlessEqual( config.get( name, 'crc8' ),       sensor.crc8 )
            self.failUnlessEqual( family,                           sensor.family )
            self.failUnlessEqual( id,                               sensor.id )
            self.failUnlessEqual( '1',                              sensor.present )
            self.failUnlessEqual( config.get( name.split( '/' )[ -1 ], 'type' ), sensor.type )

            self.failUnlessEqual( hasattr( sensor, 'PIO_0' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_1' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_2' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_3' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_4' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_5' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_6' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_7' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_ALL' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'PIO_BYTE' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'latch_0' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_1' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_2' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_3' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_4' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_5' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_6' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_7' ),     True )
            self.failUnlessEqual( hasattr( sensor, 'latch_ALL' ),   True )
            self.failUnlessEqual( hasattr( sensor, 'latch_BYTE' ),  True )
            self.failUnlessEqual( hasattr( sensor, 'power' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'reset' ),       True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_0' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_1' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_2' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_3' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_4' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_5' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_6' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_7' ),    True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_ALL' ),  True )
            self.failUnlessEqual( hasattr( sensor, 'sensed_BYTE' ), True )
            self.failUnlessEqual( hasattr( sensor, 'set_alarm' ),   True )
            self.failUnlessEqual( hasattr( sensor, 'strobe' ),      True )


    def testSet( self ):
        self.failIfEqual( len( sensors ), 0 )
        for name in sensors:
            sensor = ow.Sensor( name )

            sensor.PIO_0 = '0'
            #print sensor._path + '/PIO.0'
            self.failUnlessEqual( ow._OW.get( sensor._path + '/PIO.0' ), '0' )
            self.failUnlessEqual( sensor.PIO_0, '0' )
            self.failUnlessEqual( 'PIO_0' in dir( sensor ), False )

            sensor.PIO_0 = '1'
            self.failUnlessEqual( sensor.PIO_0, '1' )
            self.failUnlessEqual( 'PIO_0' in dir( sensor ), False )


            sensor.PIO_1 = '0'
            self.failUnlessEqual( sensor.PIO_1, '0' )
            self.failUnlessEqual( 'PIO_1' in dir( sensor ), False )

            sensor.PIO_1 = '1'
            self.failUnlessEqual( sensor.PIO_1, '1' )
            self.failUnlessEqual( 'PIO_1' in dir( sensor ), False )


            sensor.PIO_2 = '0'
            self.failUnlessEqual( sensor.PIO_2, '0' )
            self.failUnlessEqual( 'PIO_2' in dir( sensor ), False )

            sensor.PIO_2 = '1'
            self.failUnlessEqual( sensor.PIO_2, '1' )
            self.failUnlessEqual( 'PIO_2' in dir( sensor ), False )


            sensor.PIO_3 = '0'
            self.failUnlessEqual( sensor.PIO_3, '0' )
            self.failUnlessEqual( 'PIO_3' in dir( sensor ), False )

            sensor.PIO_3 = '1'
            self.failUnlessEqual( sensor.PIO_3, '1' )
            self.failUnlessEqual( 'PIO_3' in dir( sensor ), False )


            sensor.PIO_4 = '0'
            self.failUnlessEqual( sensor.PIO_4, '0' )
            self.failUnlessEqual( 'PIO_4' in dir( sensor ), False )

            sensor.PIO_4 = '1'
            self.failUnlessEqual( sensor.PIO_4, '1' )
            self.failUnlessEqual( 'PIO_4' in dir( sensor ), False )


            sensor.PIO_5 = '0'
            self.failUnlessEqual( sensor.PIO_5, '0' )
            self.failUnlessEqual( 'PIO_5' in dir( sensor ), False )

            sensor.PIO_5 = '1'
            self.failUnlessEqual( sensor.PIO_5, '1' )
            self.failUnlessEqual( 'PIO_5' in dir( sensor ), False )


            sensor.PIO_6 = '0'
            self.failUnlessEqual( sensor.PIO_6, '0' )
            self.failUnlessEqual( 'PIO_6' in dir( sensor ), False )

            sensor.PIO_6 = '1'
            self.failUnlessEqual( sensor.PIO_6, '1' )
            self.failUnlessEqual( 'PIO_6' in dir( sensor ), False )


            sensor.PIO_7 = '0'
            self.failUnlessEqual( sensor.PIO_7, '0' )
            self.failUnlessEqual( 'PIO_7' in dir( sensor ), False )

            sensor.PIO_7 = '1'
            self.failUnlessEqual( sensor.PIO_7, '1' )
            self.failUnlessEqual( 'PIO_7' in dir( sensor ), False )



def Suite( ):
    return unittest.makeSuite( DS2408, 'test' )


if __name__ == "__main__":
    if len( sys.argv ) > 1:
        unittest.main( )
    else:
        unittest.TextTestRunner( verbosity=2 ).run( Suite( ) )
